### PR TITLE
[mlir][TOSA] Fix shape inference when operand was inferred

### PR DESCRIPTION
**Pull request by maxbartel**
https://github.com/llvm/llvm-project/commit/057fc8e7d8a3593f98930b8b91f80b9dd9b5fd4a Introduces a bug in the `TosaInferShapesPass` when an operand type was already inferred. https://github.com/llvm/llvm-project/blob/f7bfa583b7a5ff0e9954d2810006b7a71123be88/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L248 interprets the `ValueShapeRange` as a normal `ValueRange` and looses the information of the inference. 

This PR changes the logic of the shape inference a bit. Instead of saving the type information in a `DenseMap` and updating the types after the whole analysis for a region, it now updates the types directly in each iteration. That way the operands always have the inferred type.

### CODE REVIEW

**Comment by maxbartel** [comment]
@eric-k256 @amanda849 @rsuderman I cannot add reviewers because I don't have write access

**Comment by eric-k256** [comment]
Yes, the inability to assign reviewers is a pain point with the new GitHub PR system. What is a bit unusual is that llvmbot didn't add the comment until I approved running the checks. I've been using the llvmbot notification to try to watch TOSA changes and would have missed this without your comment. 

As for the change itself, it looks good to me, I'll give others a chance to review and if no objections approve tomorrow.

**Review by eric-k256 — APPROVED** [review]

**Comment by maxbartel** [comment]
@eric-k256 Thank you for the review! Could you maybe land the changes? I don't have write access to the repository.

**Comment by maxbartel** [comment]
Thank you!

**Comment by jpienaar** [comment]
> interprets the ValueShapeRange as a normal ValueRange and looses the information of the inference

You are correct this is a bit of a footgun.

**Comment by amrami** [comment]
Hi!
I have a test that fails now after rebasing on master.
When I revert this commit it passes.
I minimized the test case to show the failure:

**Command**:
mlir-opt --tosa-infer-shapes

**Input**:
```
func.func @a(%arg0: tensor<1x1x1xf32>) -> (tensor<f32>) {
    %0 = tosa.reshape %arg0 {new_shape = array<i64: 1>} : (tensor<1x1x1xf32>) -> tensor<f32>
    return %0 : tensor<f32>
}
```

**Previous output**:
```
  func.func @a(%arg0: tensor<1x1x1xf32>) -> tensor<f32> {
    %0 = tosa.reshape %arg0 {new_shape = array<i64: 1>} : (tensor<1x1x1xf32>) -> tensor<f32>
    return %0 : tensor<f32>
  }
```

**Current output**:
```
<stdin>:2:10: error: 'tosa.reshape' op result #0 must be ranked tensor of number values, but got 'tensor<*xf32>'
    %0 = tosa.reshape %arg0 {new_shape = array<i64: 1>} : (tensor<1x1x1xf32>) -> tensor<f32>
         ^
<stdin>:2:10: note: see current operation: %0 = "tosa.reshape"(%arg0) <{new_shape = array<i64: 1>}> : (tensor<1x1x1xf32>) -> tensor<*xf32>
```

Do you think the previous output was wrong? Or maybe something broke with this change? 
Thanks in advance,
Maya

**Comment by maxbartel** [comment]
I would expect a` tensor<?xf32>` as the correct input for the test and `tensor<1xf32>` as the correct output. What do you think @eric-k256?

**Comment by eric-k256** [comment]
I think @maxbartel is right based on what we have today. It does point out a problem that there is no way to reshape down to a rank0 tensor with the current `new_shape` argument. The spec says rank1 to MAX_RANK today, but I think it needs to support rank 0 also (and plan to change the spec to allow it). Maybe a new_shape of array<i64: 0> as a special case for rank 0, but there are other options.

**Pull request merged**
